### PR TITLE
Preliminary support for NOC

### PR DESCRIPTION
**Pull request by clavin-xlnx**

### CODE REVIEW

File: src/com/xilinx/rapidwright/design/noc/ComponentType.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        map = new HashMap<>();
+        for(ComponentType e : values()) {
+            map.put(e.toString(), e);
+        }
+    }
+    
+    public static ComponentType stringToValue(String s) {
+        return map.get(s);
+	}
+}
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/design/noc/ComponentType.java:59]:
```suggestion
}

```

File: src/com/xilinx/rapidwright/design/noc/TrafficClass.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            }
+        }
+        return map.get(s);
+	}
+	
+	@Override
+	public String toString() {
+		return tc;
+	}
+}
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/design/noc/TrafficClass.java:58]:
```suggestion
}

```

File: src/com/xilinx/rapidwright/design/noc/ChannelType.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Enumerates the four major channel types that are used in a NOC Connection
+ */
+public enum ChannelType {
+	READ("READ"),
+	READ_REQUEST("READ_REQ"),
+	WRITE("WRITE"),
+	WRITE_RESPONSE("WRITE_RESP");
+
+    private static Map<String,ChannelType> map;
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/design/noc/ChannelType.java:36]: Consider populating in static block (like in `ComponentType`)?

File: src/com/xilinx/rapidwright/design/noc/CommunicationType.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Enumerates AXI traffic communication types.
+ */
+public enum CommunicationType {
+	MEMORY_MAPPED_FULL("MM_ReadWrite"),
+	MEMORY_MAPPED_READ("MM_ReadOnly"),
+	MEMORY_MAPPED_WRITE("MM_WriteOnly"),
+	STREAM("STRM");
+    
+    private static Map<String,CommunicationType> map;
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/design/noc/CommunicationType.java:36]: Consider populating in static block?

File: src/com/xilinx/rapidwright/design/noc/ProtocolType.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.Map;
+
+/**
+ * Enumerates AXI protocols 
+ */
+public enum ProtocolType {
+	AXI_MEMORY_MAPPED("AXI_MM"),
+	AXI_STREAM("AXI_STRM");
+    
+    private static Map<String,ProtocolType> map;
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/design/noc/ProtocolType.java:34]: Consider populating in static block?

File: src/com/xilinx/rapidwright/design/noc/TrafficClass.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Enumerates traffic classification on a given connection
+ */
+public enum TrafficClass {
+	LOW_LATENCY("LL"),
+	BEST_EFFORT("BE"),
+	ISOCHRONOUS("ISOC"),
+	BANDWIDTH("BW");
+    
+    private static Map<String,TrafficClass> map;
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/design/noc/TrafficClass.java:36]: Consider populating in static block?

File: src/com/xilinx/rapidwright/router/RouteNode.java [hunk 7]
```diff
@@ -334,22 +335,43 @@ public RouteNode getBaseWire(){
 		return new RouteNode(n);
 	}
 	
-
+	public ArrayList<PIP> getPIPsForwardToSink(){
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/router/RouteNode.java:338]: This doesn't seem very NOC related?

File: src/com/xilinx/rapidwright/router/RouteNode.java [hunk 8]
```diff
@@ -334,22 +335,43 @@ public RouteNode getBaseWire(){
...
-						break;
+			for(Wire parentWire : Arrays.asList(curr.parent.getWiresInNode())) {
+				if(!parentWire.getTile().equals(curr.getTile()))
+					continue;
+				for(Wire w1 : curr.tile.getWireConnections(parentWire.getWireIndex())){
+					if(w1.getWireIndex() == curr.wire){
+						if(w1.isEndPIPWire()){
+							pips.add(new PIP(curr.tile, parentWire.getWireIndex(), curr.wire, w1.getPIPType()));
+							break;
+						}
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/router/RouteNode.java:371]: This doesn't seem very NOC related?

File: src/com/xilinx/rapidwright/design/noc/ChannelType.java (hunk 3 above)
Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/design/noc/ChannelType.java:36]: Made all enums consistent.

File: src/com/xilinx/rapidwright/router/RouteNode.java (hunk 7 above)
Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/router/RouteNode.java:338]: Most of this code was aimed at building out relocate-ability for the NOC.  I'm assuming some additional support methods were needed and I think that is why these are here.

**Pull request merged**